### PR TITLE
Tasks/y22yo0q9/base controller

### PR DESCRIPTION
**Pull request by petertretyakov**
**Review topic**
[Base Controller](https://trello.com/c/Y22yo0q9/33-base-controller)

**Review description**
* Created `BaseController` with default 5 CRUD methods (for update there is a separate method to add it to Routes)
* Created `UserController` (conforms to Base one)
* Created `ClientController` (doesn't conforms to Base, since `Client` model is nested)
  * `client` route is `/user/:user_id/client/:client_id`
* Updated @smejl `User` and `Client` model
  * **USER**
    * Removed Identifier as it is automatically generated by Model
    * Removed ID from inits
    * Changed init(JSON) to allow initialization without all fields (only name and surname required)
  * **CLIENT**
    * Removed Identifier
    * Removed id from inits
    * Added init method from Request (it takes pushToken from JSON and userId from Request parameters)
    * Added conformance to Updateable

**Pay attention to**
Since we have discussed `BaseController` it is practically the same. Please look at `ClientController` and `Client` init from `Request`, there are a lot of unwraps there. Also check `validClient` method, which returns `BadRequest` if requested `client` is not `user`'s `client`.

**Connected tasks**
Looks like [Some User Methods](https://trello.com/c/i0ORejF4/21-some-user-methods) is half done also. `givespeech` will go to separate controller and `update photo` requires file management, as I understand correctly.

### CODE REVIEW

**Comment by ditansu** [comment]
@smejl please take in review this PR

File: Sources/App/Controllers/BaseController.swift [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+enum Method {
+  case index
+  case show
+  case create
+  case destroy
+}
+
+extension BaseController {
+  
+  func addRoutes(to droplet: Droplet, namespace: String? = nil, only methods: [Method] = [.index, .show, .create, .destroy]) {
```
Comment (smejl) [inline Sources/App/Controllers/BaseController.swift:17]: Let's do the default methods only index and show.

File: Sources/App/Controllers/ClientController.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
+struct ClientController {
```
Comment (smejl) [inline Sources/App/Controllers/ClientController.swift:1]: Why struct?

File: Sources/App/Controllers/ClientController.swift [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  
+  func addRoutes(to droplet: Droplet) {
+    droplet.get("user", User.parameter, "client", handler: index)
```
Comment (smejl) [inline Sources/App/Controllers/ClientController.swift:4]: We can use associated type for fix this case. For example: create BaseChildController, which should have parent as BaseController.

**Comment by ditansu** [comment]
@petertretyakov, please, to define the deadline to close reviewer's questions.

File: Sources/App/Controllers/ClientController.swift (hunk 2 above)
Comment (petertretyakov) [reply Sources/App/Controllers/ClientController.swift:1]: Why we can need `class` here? No inheritance, no instance modifying. But I can make `class` if you wish, not a problem ))

File: Sources/App/Controllers/BaseController.swift (hunk 1 above)
Comment (petertretyakov) [reply Sources/App/Controllers/BaseController.swift:17]: Fixed

**Comment by ditansu** [comment]
@petertretyakov close?

**Pull request closed without merging**